// angular2-presta/esm2020/lib/angular2-presta.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { HttpClientModule } from "@angular/common/http";
import { Angular2PrestaImageComponent } from "./angular2-presta-image.component";
import { Angular2PrestaService } from "./angular2-presta.service";
import * as i0 from "@angular/core";
export class Angular2PrestaModule {
    static forRoot(prestaConfiguration) {
        return {
            ngModule: Angular2PrestaModule,
            providers: [
                Angular2PrestaService,
                { provide: "prestaConfiguration", useValue: prestaConfiguration },
            ],
        };
    }
}
Angular2PrestaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.0", ngImport: i0, type: Angular2PrestaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
Angular2PrestaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.1.0", ngImport: i0, type: Angular2PrestaModule, declarations: [Angular2PrestaImageComponent], imports: [CommonModule, HttpClientModule], exports: [Angular2PrestaImageComponent] });
Angular2PrestaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.1.0", ngImport: i0, type: Angular2PrestaModule, imports: [CommonModule, HttpClientModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.0", ngImport: i0, type: Angular2PrestaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, HttpClientModule],
                    declarations: [Angular2PrestaImageComponent],
                    exports: [Angular2PrestaImageComponent],
                }]
        }] });
